File: shared/diagram-renderer/src/prepare/graph.ts
import { isOverviewVisualElementType, normalizeEdgeKind } from "../graph-normalization";
import { isDefinitionKind, isReferenceKind } from "../node-notation";
import type { PreparedNode, PreparedView, UnknownRecord, VisualizationPayload } from "./types";
import { asArray, asRecord, asString, elementTypeOf, isPackage, isSyntheticPackage, nodeUri } from "./util";

function isGeneralViewDiagramNode(node: UnknownRecord): boolean {
  if (isSyntheticPackage(node)) {
    return false;
  }
  return isOverviewVisualElementType(elementTypeOf(node));
}

function buildGeneralPackageContainerGroups(nodes: PreparedNode[]): UnknownRecord[] {
  const byPackage = new Map<string, string[]>();
  for (const node of nodes) {
    const qn = asString(asRecord(node.attributes).qualifiedName);
    const sep = qn.indexOf("::");
    if (sep <= 0) continue;
    const pkg = qn.slice(0, sep);
    const members = byPackage.get(pkg) ?? [];
    members.push(node.id);
    byPackage.set(pkg, members);
  }
  if (byPackage.size < 2) return [];
  return [...byPackage.entries()].map(([name, memberIds]) => ({
    id: `package:${name}`,
    name,
    memberIds,
  }));
}

export function prepareGraph(graphInput: unknown, visualization: VisualizationPayload): PreparedView {
  const graph = asRecord(graphInput);
  const rawNodes = asArray(graph.nodes).map(asRecord);
  const sourceNodes = rawNodes.filter((node) => isGeneralViewDiagramNode(node));
  const nodeIds = new Set(sourceNodes.map((node) => asString(node.id)));
  const nodes = sourceNodes.map((node) => ({
    id: asString(node.id),
    label: asString(node.name ?? node.qualifiedName ?? node.id, "Unnamed"),
    kind: elementTypeOf(node) || "Element",
    sourcePath: asString(node.sourcePath ?? node.source_path) || null,
    uri: nodeUri(node),
    range: (node.range as { start?: { line?: number } } | null | undefined) ?? null,
    attributes: {
      ...asRecord(node.attributes),
      qualifiedName: asString(node.qualifiedName ?? asRecord(node.attributes).qualifiedName),
      isPackage: isPackage(node),
      isDefinition: isDefinitionKind(asString(node.type ?? node.element_type, "")),
      isReference: isReferenceKind(asString(node.type ?? node.element_type, "")),
    },
  }));
  const edges = asArray(graph.edges)
    .map(asRecord)
    .filter((edge) => nodeIds.has(asString(edge.source)) && nodeIds.has(asString(edge.target)))
    .map((edge, index) => {
      const relationType = asString(edge.type ?? edge.rel_type ?? edge.relationType ?? edge.name, "");
      const label = asString(edge.name ?? edge.label ?? edge.type ?? edge.rel_type, "");
      return {
        id: asString(edge.id, `edge-${index}`),
        source: asString(edge.source),
        target: asString(edge.target),
        label,
        edgeKind: normalizeEdgeKind(relationType),
        attributes: {
          ...asRecord(edge.attributes),
          relationType: normalizeEdgeKind(relationType),
        },
      };
    });
  const packageContainerGroups = buildGeneralPackageContainerGroups(nodes);
  return {
    title: visualization?.selectedViewName || "SysML View",
    view: visualization?.view || "general-view",
    nodes,
    edges,
    meta: packageContainerGroups.length > 0 ? { packageContainerGroups } : undefined,
  };
}

File: crates/sysml_model/src/semantic/prepared_view/preparers/graph.rs
//! General and graph-backed diagram prepared views.

use std::collections::{HashMap, HashSet};

use serde_json::{json, Value};

use crate::semantic::dto::{GraphEdgeDto, GraphNodeDto, SysmlGraphDto, SysmlVisualizationResultDto};
use crate::semantic::prepared_view::dto::{PreparedEdgeDto, PreparedNodeDto, PreparedViewDto};
use crate::semantic::prepared_view::graph_norm::{
    is_definition_kind, is_overview_visual_element_type, is_reference_kind, normalize_edge_kind,
};

fn is_synthetic_package(node: &GraphNodeDto) -> bool {
    node.attributes
        .get("syntheticPackage")
        .and_then(|value| value.as_bool())
        .unwrap_or(false)
}

fn build_package_container_groups(nodes: &[PreparedNodeDto]) -> Vec<Value> {
    let mut by_package: HashMap<String, Vec<String>> = HashMap::new();
    for node in nodes {
        let Some(attrs) = node.attributes.as_ref() else {
            continue;
        };
        let Some(qn) = attrs.get("qualifiedName").and_then(|v| v.as_str()) else {
            continue;
        };
        let Some((pkg, _)) = qn.split_once("::") else {
            continue;
        };
        by_package
            .entry(pkg.to_string())
            .or_default()
            .push(node.id.clone());
    }
    if by_package.len() < 2 {
        return Vec::new();
    }
    by_package
        .into_iter()
        .map(|(name, member_ids)| {
            json!({
                "id": format!("package:{name}"),
                "name": name,
                "memberIds": member_ids,
            })
        })
        .collect()
}

pub fn prepare_graph_prepared_view(response: &SysmlVisualizationResultDto) -> PreparedViewDto {
    let graph = response
        .general_view_graph
        .as_ref()
        .or(response.graph.as_ref())
        .cloned()
        .unwrap_or(SysmlGraphDto {
            nodes: Vec::new(),
            edges: Vec::new(),
        });
    prepare_graph_from_dto(&graph, response)
}

pub fn prepare_graph_from_dto(
    graph: &SysmlGraphDto,
    response: &SysmlVisualizationResultDto,
) -> PreparedViewDto {
    let source_nodes: Vec<&GraphNodeDto> = graph
        .nodes
        .iter()
        .filter(|node| !is_synthetic_package(node) && is_overview_visual_element_type(&node.element_type))
        .collect();
    let node_ids: HashSet<String> = source_nodes.iter().map(|node| node.id.clone()).collect();
    let nodes: Vec<PreparedNodeDto> = source_nodes
        .iter()
        .map(|node| PreparedNodeDto {
            id: node.id.clone(),
            label: if node.name.is_empty() {
                node.id.clone()
            } else {
                node.name.clone()
            },
            kind: node.element_type.clone(),
            source_path: node
                .attributes
                .get("sourcePath")
                .or_else(|| node.attributes.get("source_path"))
                .and_then(|v| v.as_str())
                .map(str::to_string),
            uri: node.uri.clone(),
            range: Some(node.range.clone()),
            attributes: Some(json!({
                "qualifiedName": node.id,
                "isPackage": node.element_type.to_lowercase().contains("package"),
                "isDefinition": is_definition_kind(&node.element_type),
                "isReference": is_reference_kind(&node.element_type),
            })),
        })
        .collect();
    let edges: Vec<PreparedEdgeDto> = graph
        .edges
        .iter()
        .filter(|edge| node_ids.contains(&edge.source) && node_ids.contains(&edge.target))
        .enumerate()
        .map(|(index, edge)| graph_edge_to_prepared(edge, index))
        .collect();
    let package_container_groups = build_package_container_groups(&nodes);
    let title = response
        .selected_view_name
        .clone()
        .unwrap_or_else(|| "SysML View".to_string());
    PreparedViewDto {
        title,
        view: response.view.clone(),
        nodes,
        edges,
        meta: if package_container_groups.is_empty() {
            None
        } else {
            Some(json!({ "packageContainerGroups": package_container_groups }))
        },
    }
}

fn graph_edge_to_prepared(edge: &GraphEdgeDto, index: usize) -> PreparedEdgeDto {
    let rel_type = edge.rel_type.clone();
    let label = edge
        .name
        .clone()
        .unwrap_or_else(|| rel_type.clone());
    PreparedEdgeDto {
        id: format!("edge-{index}"),
        source: edge.source.clone(),
        target: edge.target.clone(),
        label,
        edge_kind: Some(normalize_edge_kind(&rel_type)),
        attributes: Some(json!({ "relationType": normalize_edge_kind(&rel_type) })),
    }
}
